src/x86/x86filter.c

```diff
@@ -0,0 +1,180 @@
+/*Daala video codec
+Copyright (c) 2015 Daala project contributors.  All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions are met:
+
+- Redistributions of source code must retain the above copyright notice, this
+  list of conditions and the following disclaimer.
+
+- Redistributions in binary form must reproduce the above copyright notice,
+  this list of conditions and the following disclaimer in the documentation
+  and/or other materials provided with the distribution.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS “AS IS”
+AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE
+FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.*/
+
+#if defined(HAVE_CONFIG_H)
+# include "config.h"
+#endif
+
+#include "x86enc.h"
+#include "cpu.h"
+
+#if defined(OD_X86ASM)
+#include <stdio.h>
+#include <emmintrin.h>
+#include <pmmintrin.h>
+#include <tmmintrin.h>
+#include "../filter.h"
+
+#if defined(OD_CHECKASM)
+void od_filter_dering_direction_check(int16_t *y, int ystride, int16_t *in,
+ int bstride, int log_n, int threshold, int dir) {
+  int16_t dst[OD_BSIZE_MAX*OD_BSIZE_MAX];
+  od_filter_dering_direction_c(dst, OD_BSIZE_MAX, in, bstride, log_n,
+   threshold, dir);
+  int i;
+  int j;
+  int failed;
+  int n = 1 << log_n;
+  failed = 0;
+  for (i = 0; i < n; i++) {
+    for (j = 0; j < n; j++) {
+      if (dst[i*OD_BSIZE_MAX + j] != y[i*ystride + j]) {
+        fprintf(stderr,"ASM mismatch: 0x%02X!=0x%02X @ (%2i,%2i)\n",
+         dst[i*OD_BSIZE_MAX + j],y[i*ystride + j],i,j);
+        failed = 1;
+      }
+    }
+  }
+  if (failed) {
+    fprintf(stderr, "od_filter_dering_direction check failed.\n");
+  }
+}
+#endif
+
+typedef void (*od_filter_dering_direction_fixed_func)(int16_t *y, int ystride,
+ int16_t *in, int bstride, int threshold, int dir);
+
+void od_filter_dering_direction_4x4(int16_t *y, int ystride, int16_t *in,
+ int bstride, int threshold, int dir) {
+  int i;
+  int k;
+  static const int taps[3] = {3, 2, 2};
+  int offset[3];
+  __m128i sum;
+  __m128i p0;
+  __m128i cmp0;
+  __m128i p1;
+  __m128i cmp1;
+  __m128i row;
+  __m128i res;
+  od_filter_dering_direction_offsets(offset, dir, bstride);
```

Comment:
bstride is a constant. dir is only ever a small set of values. You could use a constant array.

bstride = (OD_BSIZE_MAX + 2*OD_FILT_BORDER). Will it stay that way?